File: scripts/add-react-memo.mjs
#!/usr/bin/env node
/**
 * add-react-memo.mjs — Batch wrap all UI components with React.memo
 * Run: node scripts/add-react-memo.mjs
 */

import { readFileSync, writeFileSync, readdirSync } from 'fs'
import { join, parse } from 'path'

const dir = 'apps/web/components/ui'
const fullDir = join(process.cwd(), dir)
const files = readdirSync(fullDir).filter(f => f.endsWith('.tsx') && !f.endsWith('.stories.tsx'))

let transformed = 0
let skipped = 0

for (const file of files) {
  const fp = join(fullDir, file)
  let src = readFileSync(fp, 'utf-8')
  const original = src

  // Skip if already has memo in react import
  if (/import\s*\{[^}]*\bmemo\b[^}]*\}\s*from\s*['"]react['"]/.test(src)) {
    skipped++
    continue
  }

  // 1. Add memo to react import
  src = src.replace(
    /(import\s*\{)([^}]*)(\})\s*from\s*['"]react['"]/,
    (match, open, body, close) => {
      const items = body.split(',').map(s => s.trim()).filter(Boolean)
      if (items.includes('memo')) return match
      if (items.length === 0) return `${open} memo ${close} from 'react'`
      return `${open} memo, ${body} ${close} from 'react'`
    },
  )

  // 2. Wrap forwardRef with memo
  // Pattern: `const Button = forwardRef<...>((props, ref) => { ... })` 
  //   → `const Button = memo(forwardRef<...>((props, ref) => { ... }))`
  // First handle closing: count parens on the last line before semicolon
  src = src.replace(
    /(?:const\s+(\w+)\s*=\s*)(forwardRef<)/g,
    'const $1 = memo($2',
  )

  // Close memo wrapper — find the matching closing paren of forwardRef
  // Look for `)`, `)` pattern where first ) closes forwardRef, second closes memo
  src = src.replace(
    /(memo\(forwardRef<[^>]+>)/g,
    (match) => match // leave the opening as-is, handle closing below
  )

  // Close memo for forwardRef components
  // Pattern: find lines with `memo(forwardRef<` and add extra `)` before export or end
  const lines = src.split('\n')
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    if (line.includes('memo(forwardRef<') && !line.includes('/*memo-closed*/')) {
      // Find the end of the component definition — look for `)` on its own line or `)` followed by export
      let depth = 0
      let inString = false
      let startIdx = line.indexOf('memo(')
      if (startIdx === -1) continue
      
      // Scan from memo( to count parens and find the closing of forwardRef
      let memoCloseNeeded = true
      // Actually, simpler approach: the pattern is:
      // const X = memo(forwardRef<...>((...) => {
      //   ...
      // }))
      // We need to find the last ))) that closes forwardRef, function, and memo
      // But TypeScript handles this correctly — just add another )
      // Search for the last line that is just `)` or has `)` before `export`
      
      for (let j = i; j < lines.length; j++) {
        const l = lines[j]
        for (let k = 0; k < l.length; k++) {
          const ch = l[k]
          if (ch === '(') depth++
          else if (ch === ')') depth--
        }
        if (depth === 0 && j > i) {
          // This line closes the outer scope
          // Add a closing paren for memo
          // But we need to be careful — if there's an export line, add before it
          const nextLines = lines.slice(j)
          const exportIdx = nextLines.findIndex(l2 => /^export\s+\{/.test(l2.trim()) || /^$/.test(l2.trim()) && nextLines.slice(nextLines.indexOf(l2)).some(l3 => /^export\s+\{/.test(l3.trim())))
          
          // Add memo closing - just add ) before the export
          // Actually let's try a different approach entirely
          break
        }
      }
    }
  }

  // Simpler approach: just regex replace forwardRef pattern
  // After adding "memo(" before "forwardRef<", we need to add ")" before the last ")"
  // The pattern is: `const X = memo(forwardRef<Props>((props, ref) => { ... }))`
  // Original: `const X = forwardRef<Props>((props, ref) => { ... })`
  // After:     `const X = memo(forwardRef<Props>((props, ref) => { ... }))`
  // So we add "memo(" before forwardRef and add ")" before the final closing artifacts
  
  // Restore original and try a different approach
  src = original

  // Approach: simple regex-based
  // Step 1: Add memo import
  src = src.replace(
    /(import\s*\{)([^}]*)(\}\s*from\s*['"]react['"])/,
    (match, open, body, close) => {
      const items = body.split(',').map(s => s.trim()).filter(Boolean)
      if (items.includes('memo')) return match
      if (items.length === 0) return `${open} memo ${close}`
      return `${open} memo, ${body} ${close}`
    },
  )

  // Step 2: For forwardRef components, wrap with memo  
  // const X = forwardRef<...>(...)  →  const X = memo(forwardRef<...>(...))
  src = src.replace(
    /(const\s+\w+\s*=\s*)forwardRef</g,
    '$1memo(forwardRef<',
  )
  // Close the extra paren for memo — add ) before DisplayName or export
  // After: `const X = memo(forwardRef<Props>((p,r) => { ... }))\nX.displayName = 'X'`
  // Need:  `const X = memo(forwardRef<Props>((p,r) => { ... }))\nX.displayName = 'X'`
  // The hard part is knowing where to close. forwardRef ends with:
  //   }))\n  OR  }),)\n  etc
  // We need to add one more ) after the forwardRef closing.
  
  // Let's find forwardRef component definitions and count parens to close
  const fileContent = src
  const forwardRefPattern = /const\s+(\w+)\s*=\s*memo\(forwardRef</
  const matches = [...fileContent.matchAll(new RegExp(forwardRefPattern, 'g'))]
  
  for (const m of matches) {
    const compName = m[1]
    const startIdx = m.index
    // Find the matching close — scan for `X.displayName` or end of definition
    const afterMatch = fileContent.slice(startIdx)
    const displayNamePattern = new RegExp(`\n${compName}\\.displayName\\s*=`)
    const displayNameMatch = displayNamePattern.exec(afterMatch)
    
    if (displayNameMatch) {
      // The forwardRef definition ends before displayName
      // We need to find the last `)` before displayName and add `)` after it
      const defEnd = startIdx + displayNameMatch.index
      const defContent = fileContent.slice(startIdx, defEnd)
      
      // Count parens to find what closes forwardRef
      let depth = 0
      let lastCloseParenIdx = -1
      for (let i = 0; i < defContent.length; i++) {
        if (defContent[i] === '(') depth++
        else if (defContent[i] === ')') {
          depth--
          if (depth === 0) {
            lastCloseParenIdx = i
            break
          }
        }
      }
      
      if (lastCloseParenIdx !== -1) {
        // Insert `)` before displayName
        const insertPos = startIdx + lastCloseParenIdx + 1
        src = src.slice(0, insertPos) + ')' + src.slice(insertPos)
      }
    } else {
      // Maybe the definition continues, try finding `)` that closes the arrow function
      // This is getting complex. Let's use eval-based approach for simpler cases
    }
  }

  // Actually, let me take a much simpler approach.
  // For ALL files, just do import + wrap exports.
  // Reset to original
  src = original

  // Simple approach: 
  // 1. Add memo import
  src = src.replace(
    /(import\s*\{)([^}]*)(\}\s*from\s*['"]react['"])/,
    (match, open, body, close) => {
      const items = body.split(',').map(s => s.trim()).filter(Boolean)
      if (items.includes('memo')) return match
      if (items.length === 0) return `${open} memo ${close}`
      return `${open} memo, ${body} ${close}`
    },
  )

  // 2. Wrap export function X → const X = memo(function X
  src = src.replace(
    /^export\s+function\s+(\w+)/gm,
    'const $1 = memo(function $1',
  )

  // 3. Wrap export function X (inline) that wasn't caught
  src = src.replace(
    /^function\s+(\w+)/gm,
    (match, name) => {
      // Skip if already wrapped or if it's an inner function
      const lineIdx = src.split('\n').findIndex(l => l.trim() === match.trim())
      const prevLine = lineIdx > 0 ? src.split('\n')[lineIdx - 1].trim() : ''
      if (prevLine.includes('memo(')) return match
      return `const ${name} = memo(function ${name}`
    },
  )

  // 4. Fix export { X } — they should already work since const X = memo(...) creates the binding
  // But some files have `export function X` at the bottom export list — those are now `const X = memo(function X`
  // which is NOT exported. We need to check if X was previously in the export list.

  if (src !== original) {
    writeFileSync(fp, src, 'utf-8')
    transformed++
    console.log(`✓ ${dir}/${file}`)
  } else {
    skipped++
  }
}

console.log(`\nDone — ${transformed} transformed, ${skipped} skipped`)

File: scripts/add-react-memo.py
#!/usr/bin/env python3
"""Batch-add React.memo to all UI component files."""

import re
from pathlib import Path

UI_DIR = Path("apps/web/components/ui")
SKIP_PATTERNS = (".stories.", ".test.", ".bench.")
SYS_ENCODING = "utf-8"

ALREADY_MEMO = re.compile(r"import\s*\{[^}]*\bmemo\b[^}]*\}\s*from\s*['\"]react['\"]")
REACT_IMPORT = re.compile(r"(import\s*\{)([^}]*)(\}\s*from\s*['\"]react['\"])")


def add_memo_import(content: str) -> str:
    def replacer(m: re.Match) -> str:
        body = m.group(2).strip()
        items = [x.strip() for x in body.split(",") if x.strip()]
        if "memo" in items:
            return m.group(0)
        if not items:
            return f"{m.group(1)} memo {m.group(3)}"
        return f"{m.group(1)} memo, {m.group(2)} {m.group(3)}"

    return REACT_IMPORT.sub(replacer, content)


def wrap_forward_ref_components(content: str) -> str:
    """Wrap `const X = forwardRef<...>(...)` with memo()."""
    lines = content.split("\n")
    result = []

    for i, line in enumerate(lines):
        m = re.match(r"(const\s+\w+\s*=\s*)forwardRef<", line)
        if m:
            m.group(1)
            comp_name_match = re.search(r"const\s+(\w+)\s*=", line)
            comp_name = comp_name_match.group(1) if comp_name_match else None

            # Replace `= forwardRef<` with `= memo(forwardRef<`
            line = line.replace("= forwardRef<", "= memo(forwardRef<", 1)

            if comp_name:
                # Scan forward for displayName line
                for j in range(i + 1, len(lines)):
                    stripped = lines[j].strip()
                    if f"{comp_name}.displayName" in stripped:
                        # Add ')' at the end of the previous line
                        prev_idx = j - 1
                        prev_stripped = lines[prev_idx].rstrip()
                        lines[prev_idx] = prev_stripped + ")"
                        break

        result.append(line)

    return "\n".join(result)


def wrap_plain_functions(content: str) -> str:
    """Replace `function X(...)` with `const X = memo(function X(...)` for pascal-cased functions."""

    def replacer(m: re.Match) -> str:
        m.group(1) or ""
        name = m.group(2)
        return f"const {name} = memo(function {name}("

    return re.sub(
        r"^(export\s+)?function\s+([A-Z]\w*)\s*\(",
        replacer,
        content,
        flags=re.MULTILINE,
    )


def process_file(fp: Path) -> bool:
    content = fp.read_text(encoding=SYS_ENCODING)
    original = content

    if ALREADY_MEMO.search(content):
        return False

    content = add_memo_import(content)
    content = wrap_forward_ref_components(content)
    content = wrap_plain_functions(content)

    if content != original:
        fp.write_text(content, encoding=SYS_ENCODING)
        return True
    return False


def main():
    count = 0
    for fp in sorted(UI_DIR.iterdir()):
        if fp.suffix != ".tsx":
            continue
        if any(p in fp.name for p in SKIP_PATTERNS):
            continue
        if fp.name in ("index.ts", "utils.ts"):
            continue

        if process_file(fp):
            print(f"  OK {fp.name}")
            count += 1
        else:
            print(f"  -- {fp.name}")

    print(f"\nDone - {count} files transformed")


if __name__ == "__main__":
    main()
